rfd/0169-auto-updates-linux-agents.md

```diff
@@ -91,45 +89,145 @@ spec:
   # cluster level. Only turn agent automatic updates off if self-managed
   # agent updates are in place.
   agent_auto_update: true|false
-  # agent_update_hour sets the hour in UTC at which clients should update their agents.
-  agent_update_hour: 0-23
-  # agent_update_jitter_seconds sets a duration in which the upgrade will occur after the hour.
-  # The agent upgrader will pick a random time within this duration to wait to upgrade.
-  agent_update_jitter_seconds: 0-3600
-  
-  [...]
+
+  # agent_auto_update_groups contains both "regular" and "critical" schedules.
+  # The schedule used is determined by the agent_version_schedule associated
+  # with the version in autoupdate_version.
+  # Groups are not configurable with the "immediate" schedule.
+  agent_auto_update_groups:
+    # schedule is "regular" or "critical"
+    regular:
+    - name: staging-group
+      # agent_selection defines which agents are included in the group.
+      agent_selection:
+        # query selects agents by resource query.
+        #  default: all connected agents
+        query: 'labels["environment"]=="staging"'
+      # days specifies the days of the week when the group may be upgraded.
+      #  default: ["*"] (all days)
+      days: [“Sun”, “Mon”, ... | "*"]
+      # start_hour specifies the hour when the group may start upgrading.
+      #  default: 0
+      start_hour: 0-23
+      # jitter_seconds specifies a maximum jitter duration after the start hour.
+      # The agent upgrader client will pick a random time within this duration to wait to upgrade.
+      #  default: 0
+      jitter_seconds: 0-60
+      # max_in_flight specifies the maximum number of agents that may be upgraded at the same time.
+      #  default: 100%
+      max_in_flight: 0-100%
+      # timeout_seconds specifies the amount of time, after the specified jitter, after which
+      # an agent upgrade will be considered timed out if the version does not change.
+      #  default: 60
+      timeout_seconds: 30-900
+      # failure_seconds specifies the amount of time after which an agent upgrade will be considered
+      # failed if the agent heartbeat stops before the upgrade is complete.
```

Comment:
Instance heartbeats happen at intervals of several minutes, and that's currently the only way for proxies (and the cluster at large) to know if a host is at a given version. How are we going to keep track of initiated and failed upgrades?

Reply:
The interval is documented at 60s, and I'm assuming that an upgraded agent will heartbeat immediately (let me know if that's incorrect).

Failed agents miss heartbeats for a configurable period of time, while timed-out agent continue to heartbeat the old version for a configurable period of time. Pre-upgrade agent health can be determined right before the upgrade is initiated for that agent.

The 60s heartbeat interval means that phased upgrades will happen relatively slowly -- I think that's okay.